# Tarea 1: Introducción, Vector Space Models, Information Retrieval y Language Models</h1>
**Procesamiento de Lenguaje Natural (CC6205-1 - Otoño 2024)**

## Tarjeta de identificación

**Nombres:**
1. Martín Bravo
2. Felipe Fierro

**Fecha límite de entrega 📆:** 10/04.

**Tiempo estimado de dedicación:** 4 horas


## Instrucciones
Bienvenid@s a la primera tarea en el curso de Natural Language Processing (NLP). Esta tarea tiene como objetivo evaluar los contenidos teóricos de las primeras semanas de clases, enfocado principalmente en **Information Retrieval (IR)**, **Vector Space Models** y **Language Models**. Si aún no has visto las clases, se recomienda visitar los links de las referencias.

La tarea consta de una parte teórica que busca evaluar conceptos vistos en clases. Seguido por una parte práctica con el fín de introducirlos a la programación en Python enfocada en NLP.

* La tarea es en **grupo** (maximo hasta 3 personas).
* La entrega es a través de u-cursos a más tardar el día estipulado arriba. No se aceptan atrasos.
* El formato de entrega es este mismo Jupyter Notebook.
* Al momento de la revisión su código será ejecutado. Por favor verifiquen que su entrega no tenga errores de compilación.
* Completar la tarjeta de identificación. Sin ella no podrá tener nota.

## Material de referencia

Diapositivas del curso 📄
    
- [Introducción al curso](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-introduction.pdf)
- [Vector Space Model / Information Retrieval](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-IR.pdf)    
- [Probabilistic Language Models](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-PLM.pdf)

Videos del curso 📺
- Introducción  [Parte 1](https://www.youtube.com/watch?v=HEKTNOttGvU)  [Parte 2](https://www.youtube.com/watch?v=P8cwnI-f-Kg)

- Information Retrieval [Parte 1](https://www.youtube.com/watch?v=FXIVClF370w&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=3) [Parte 2](https://www.youtube.com/watch?v=f8nG1EMmPZk&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=3)
- Probabilistic Language Models [Parte 1](https://www.youtube.com/watch?v=9E2jJ6kcb4Y&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=3) [Parte 2](https://www.youtube.com/watch?v=ZWqbEQXLra0&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=5) [Parte 3](https://www.youtube.com/watch?v=tsumFqwFlaA&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=6) [Parte 4](https://www.youtube.com/watch?v=s3TWdv4sqkg&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJq)

## P1. Tokenización

En el primer ejercicio veremos la dificultad de tokenizar textos no estructurados, destacando la importancia de tener librerías que realicen este trabajo.

### Pregunta 1.a (0.25 puntos)

Diseñe una función **`get_tokens()`** que reciba un texto y entregue una lista con sus tokens. Es libre de elegir la forma de tokenizar mientras no utilice librerías con tokenizadores ya implementados. Puede utilizar la librería **re** importada para trabajar símbolos. Explique su razonamiento.


**Ya que estamos trabajando el documento de forma local vamos a poner el texto directamente acá**

In [113]:
# Abrir oh_algoritmo.txt
with open('oh_algoritmo.txt', 'r') as file:
    texto = file.read()

texto

'[Letra de "¡Oh, Algoritmo!" ft. Nora Erez]\n\n[Refrán: Jorge Drexler]\n¿Quién quiere que yo quiera lo que creo que quiero?\n¿Quién quiere que yo quiera lo que creo que quiero?\n¿Quién quiere que yo quiera lo que creo que quiero?\n¿Quién quiere que yo quiera lo que creo que quiero?\n¿Quién quiere que yo quiera lo que creo que quiero?\n¿Quién quiere que yo quiera lo que creo que quiero?\n\n[Estribillo: Jorge Drexler]\nDime qué debo cantar\nOh, algoritmo\nSé que lo sabes mejor\nIncluso que yo mismo\n\n[Verso 1: Nora Erez]\nWait, what\'s that money that you spent?\nWhat\'s that sitting on your plate?\nDo you want what you\'ve been fed?\nAre you the fish or bait?\nMmm, I\'m on the top of the roof and I feel like a jail\nRather not pay the bail\nTo dangerous people with blood on their faces\nSo I\'m sharing a cell with the masses\nThe underground always strive for the main\nStreaming like Grande\'s big-ass ring\nScreaming: I\'ll write you out my will\nConscious is free, but not the will\nCo

In [114]:
import re

In [115]:
def get_tokens(text):
    

    # Reemplazar \n por espacio
    text = text.replace("\n", " ")
    # Separar palabras entre corchetes y parentesis
    text = re.sub(r'\[.*?\]', ' ', text)
    text = re.sub(r'\(.*?\)', ' ', text)
    # Separar signos de puntuación, guiones, guiones bajos, comillas y otros
    text = re.sub(r'[\-\_.,;:!?¿¡\']', ' ', text)

    # Separar palabras
    tokens = text.split()
    
    # Eliminar tokens vacíos (en caso de haber espacios adicionales)
    tokens = [token for token in tokens if token.strip()]
    
    return tokens

In [116]:
tokens = get_tokens(texto)

### Pregunta 1.b (0.25 puntos)
Explique su implementación aquí:
> La implementación consiste en todos los saltos de línea por espacios, luego los corchetes, parentesis y llaves son separados de las palabras, luego se eliminan los signos de puntuación y se separan las palabras por espacios. Finalmente se eliminan los espacios en blanco y se retorna una lista con las palabras.

Implementación con la libreria NLTK

In [117]:
from nltk.tokenize import wordpunct_tokenize
nltk_tokens = wordpunct_tokenize(texto)

Podemos verificar las diferencias entre ambos códigos, podemos ver que son muy pocas con respecto al algoritmo de NLTK

In [118]:
print(set(tokens) - set(nltk_tokens))
print(set(nltk_tokens) - set(tokens))

set()
{')', '1', '4', ':', '"¡', '¿', ',', '!"', 'Erez', "'", '-', '3', 'Refrán', '[', '2', 'Letra', 'Algoritmo', 'Pre', 'ft', 'Wow', '?', ']', 'Nora', 'Estribillo', '.', '(', 'Verso'}


### Pregunta 1.c (0.5 puntos)
¿Qué diferencias y similitudes encontrase al comparar la función de tokenización creada manualmente por ti contra la implementación de NLTK, al tokenizar la letra de la canción "Oh, algoritmo"?

> Algunas diferencias que se pueden encontrar son que la implementación manual no considera los signos de puntuación como tokens, mientras que la implementación de NLTK si lo hace. Por otro lado, ambas implementaciones consideran los saltos de línea como espacios, y eliminan los espacios en blanco al final de cada token.

## P2. Stemming y Stopwords

En esta sección debera implementar funciones de stemming y stopwords basado en lo visto en clase. En la siguiente celda tiene el corpus que usara en esta sección:

In [119]:
# Corpus en español
corpus_espanol = [
    "¿Quién quiere que yo quiera lo que creo que quiero?",
    "Dime qué debo cantar",
    "Sé que lo sabes mejor"
]

# Corpus en inglés
corpus_ingles = [
    "What's that sitting on your plate?",
    "Do you want what you've been fed?",
    "Are you the fish or bait?"
]

### Pregunta 2.a (0.5 puntos)
Implemente una función **`get_vocab()`** que extraiga los tokens de un corpus. Puede utilizar la función de la sección anterior.

In [120]:
def get_vocab(corpus):

  corpus_complete = ""
  for i in range(len(corpus)):
    corpus_complete += corpus[i] + " "
  tokens = get_tokens(corpus_complete)
  return tokens

In [121]:
# Obtenemos el vocabulario del corpus en español
vocab_espanol = get_vocab(corpus_espanol)

# Comprobamos que el vocabulario obtenido es correcto
set(['yo', 'debo', 'creo', 'Dime', 'lo', 'cantar', 'mejor', 'Sé', 'que', 'quiere', 'quiero', 'sabes', 'Quién', 'quiera', 'qué']) == set(vocab_espanol) 


True

Resultado esperado (el orden puede variar):
```
['yo', 'debo', 'creo', 'Dime', 'lo', 'cantar', 'mejor', 'Sé', 'que', 'quiere', 'quiero', 'sabes', 'Quién', 'quiera', 'qué']
```

In [122]:
# Obtenemos el vocabulario del corpus en inglés
vocab_ingles = get_vocab(corpus_ingles)

# Comprobamos que el vocabulario obtenido es correcto
set(vocab_ingles) == set(['fed', 'been', 'or', 'want', 'plate', 'the', 've', 'your', 's', 'you', 'what', 'Are', 'bait', 'What', 'fish', 'that', 'sitting', 'Do', 'on'])

True

Resultado esperado:
```
['fed', 'been', 'or', 'want', 'plate', 'the', 've', 'your', 's', 'you', 'what', 'Are', 'bait', 'What', 'fish', 'that', 'sitting', 'Do', 'on']
```

### Pregunta 2.b (0.5 puntos)
Ahora diseñe reglas que usted estime convenientes tanto de **Stemming** como de **Stopwords**. Implemente una función que reciba una lista con los elementos del vocabulario, le aplique sus reglas y devuelva el vocabulario preprocesado. Explique las reglas de stemming y elección de stopwords:

    Explique sus reglas aquí:


Primero definimos los sufijos de los que nos queremos deshacer para hacer el stemming, luego definimos las stopwords que queremos eliminar del vocabulario.

Luego pasamos todas las palabras a minúsculas eliminando las tildes, finalmente eliminamos los sufijos definidos y las stopwords.

In [123]:
def pre_processing(vocabulario, idioma):

    # Definimos los sufijos de los cuales queremos deshacernos
    stemming = {
        "espanol": ["ar", "er", "ir", "a", "e", "o", "s", "es"],
        "ingles": ["s", "es", "ed", "ing", "ly", "er", "est", "ing"]
    }

    # Definimos las palabras que queremos eliminar
    stopwords = {
        "espanol": ["yo", "lo", "que", "qué", "debo", "Sé"],
        "ingles": ["or", "the", "ve", "s", "what", "are", "Are", "that", "What", "Do", "on", "been", "you", "your", "fed"]
    }            

    # Creamos una copia del vocabulario
    new_vocab = vocabulario.copy()

    # Convertimos todas las palabras a minúsculas
    new_vocab = [palabra.lower() for palabra in new_vocab]

    # Eliminar tilde
    new_vocab = [palabra.replace("á", "a").replace("é", "e").replace("í", "i").replace("ó", "o").replace("ú", "u") for palabra in new_vocab]

    # Eliminar sufijos y stopwords
    for palabra in range(len(new_vocab)):
        for sufijo in stemming[idioma]:
            if new_vocab[palabra].endswith(sufijo):
                new_vocab[palabra] = new_vocab[palabra][:-len(sufijo)]
        if new_vocab[palabra] in stopwords[idioma]:
            new_vocab[palabra] = ""

    new_vocab = [palabra for palabra in new_vocab if palabra]

    return set(new_vocab)

In [124]:
# Aplicar preprocesamiento a los vocabularios de ejemplo con NLTK
vocab_procesado_espanol = pre_processing(vocab_espanol, 'espanol')
vocab_procesado_ingles = pre_processing(vocab_ingles, 'ingles')

# Mostrar resultados
print("Vocabulario procesado en español:", vocab_procesado_espanol, "\n")
print("Vocabulario procesado en inglés:", vocab_procesado_ingles, "\n")

Vocabulario procesado en español: {'cant', 'sabe', 'l', 'deb', 'quier', 'y', 'mejor', 'qu', 'cre', 'quien', 'dim'} 

Vocabulario procesado en inglés: {'f', 'do', 'fish', 'sitt', 'plate', 'want', 'bait'} 



## P3. Bag of Words (0.5 puntos)
Considere el siguiente corpus, donde cada elemento del arreglo representa un documento:


In [125]:
d0 = 'El pájaro come semillas'
d1 = 'El pájaro se despierta y canta'
d2 = 'El pájaro canta y come semillas'
d3 = 'El pez come y nada en el agua'
d4 = 'El pez empieza a nadar'
d5 = 'El pez come alimento'
corpus = [d0, d1, d2, d3, d4, d5]

El objetivo da las siguientes secciones es determinar cuáles de  los documentos entregados son los más similares entre sí. Para ello utilizaremos la técnica **TF-IDF**.

Como los algoritmos de Machine Learning no comprenden el texto en lenguaje natural, estos documentos deben ser convertidos a vectores numéricos. La representación más simple vista en clases es la de **Bag of Words**, método mediante el cual se cuentan las apariciones de cada palabra en cada uno de los documentos entregados.

Implemente la función **`bag_of_words()`**, que recibe como input un arreglo de documentos y devuelve un dataframe de pandas con la representación Bag of Words de los documentos entregados. En esta representación las columnas son el vocabulario y las filas representan las apariciones de cada una de las palabras en los documentos. En otras palabras, cada fila representa el BoW de un documento.

***Disclaimer: el orden de los resultados pueden variar.***


Por ejemplo para el siguiente corpus:

```
corpus = ['El perro ladra', 'El perro come']
```

Debiese entregarnos lo siguiente:


|   | el | perro | ladra | come |
|---|----|-------|------|-------|
| 0 | 1  | 1     | 1    | 0     |
| 1 | 1  | 1     | 0    | 1     |




In [126]:
import pandas as pd

Implementar función `bag_of_words()`

In [127]:
def bag_of_words(corpus):
    
    # Obtener el vocabulario del corpus
    vocab = list(set(get_vocab(corpus)))

    # Crear un DataFrame con las palabras del vocabulario como columnas
    df = pd.DataFrame(columns=vocab, index=range(len(corpus)))

    # Llenar el DataFrame con ceros
    df = df.fillna(0)

    # Recorrer los documentos del corpus
    for i in range(len(corpus)):
        # Obtener los tokens del documento
        tokens = get_tokens(corpus[i])
        # Recorrer las palabras del vocabulario
        for palabra in vocab:
            # Contar cuántas veces aparece la palabra en el documento
            df.loc[i, palabra] = tokens.count(palabra)

    return df

In [128]:
dataset_bow = bag_of_words(corpus)

dataset_bow

se  pez  El  pájaro  en  nada  alimento  y  nadar  empieza  semillas  \
0   0    0   1       1   0     0         0  0      0        0         1   
1   1    0   1       1   0     0         0  1      0        0         0   
2   0    0   1       1   0     0         0  1      0        0         1   
3   0    1   1       0   1     1         0  1      0        0         0   
4   0    1   1       0   0     0         0  0      1        1         0   
5   0    1   1       0   0     0         1  0      0        0         0   

   canta  come  despierta  el  agua  a  
0      0     1          0   0     0  0  
1      1     0          1   0     0  0  
2      1     1          0   0     0  0  
3      0     1          0   1     1  0  
4      0     0          0   0     0  1  
5      0     1          0   0     0  0

Solución esperada:

|    |   El |   pájaro |   despierta |   el |   come |   a |   nadar |   se |   en |   y |   alimento |   semillas |   pez |   empieza |   canta |   agua |   nada |
|:---|-----:|---------:|------------:|-----:|-------:|----:|--------:|-----:|-----:|----:|-----------:|-----------:|------:|----------:|--------:|-------:|-------:|
| d0 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          0 |          1 |     0 |         0 |       0 |      0 |      0 |
| d1 |    1 |        1 |           1 |    0 |      0 |   0 |       0 |    1 |    0 |   1 |          0 |          0 |     0 |         0 |       1 |      0 |      0 |
| d2 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   1 |          0 |          1 |     0 |         0 |       1 |      0 |      0 |
| d3 |    1 |        0 |           0 |    1 |      1 |   0 |       0 |    0 |    1 |   1 |          0 |          0 |     1 |         0 |       0 |      1 |      1 |
| d4 |    1 |        0 |           0 |    0 |      0 |   1 |       1 |    0 |    0 |   0 |          0 |          0 |     1 |         1 |       0 |      0 |      0 |
| d5 |    1 |        0 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          1 |          0 |     1 |         0 |       0 |      0 |      0 |

## P4. TF-IDF

### 4.a TF (0.25 puntos)

Ahora debemos usar el dataframe del ejercicio anterior para calcular la matriz de TF normalizada por la máxima frecuencia $\max_i({\text{tf}_{i,j}})$, donde
$i$ corresponde al índice de las filas (BoW) y $j$ al de las columnas (palabras). Es decir, dividir cada BoW sobre la cantidad de veces de la palabra que aparezca más veces en ese vector.


$$\text{nft}_{i,j} = \frac{\text{tf}_{i,j}}{\max_i({\text{tf}_{i,j})}}$$

Implemente la función `calc_tf(dataset_bow)`, que entrega la matriz de TF normalizada del BoW del dataset.

In [129]:
def calc_tf(df):
    
    # Crear un DataFrame para almacenar los valores de TF
    tf = pd.DataFrame(columns=df.columns, index=range(len(df)))

    # Recorrer las filas
    for i in range(len(df)):
        # Calcular el máxima de la fila
        max = df.loc[i].max()
        # Calcular el TF para cada palabra
        for palabra in df.columns:
            tf.loc[i, palabra] = df.loc[i, palabra] / max

    return tf

In [130]:
tf = calc_tf(dataset_bow)
tf

se  pez   El pájaro   en nada alimento    y nadar empieza semillas canta  \
0  0.0  0.0  1.0    1.0  0.0  0.0      0.0  0.0   0.0     0.0      1.0   0.0   
1  1.0  0.0  1.0    1.0  0.0  0.0      0.0  1.0   0.0     0.0      0.0   1.0   
2  0.0  0.0  1.0    1.0  0.0  0.0      0.0  1.0   0.0     0.0      1.0   1.0   
3  0.0  1.0  1.0    0.0  1.0  1.0      0.0  1.0   0.0     0.0      0.0   0.0   
4  0.0  1.0  1.0    0.0  0.0  0.0      0.0  0.0   1.0     1.0      0.0   0.0   
5  0.0  1.0  1.0    0.0  0.0  0.0      1.0  0.0   0.0     0.0      0.0   0.0   

  come despierta   el agua    a  
0  1.0       0.0  0.0  0.0  0.0  
1  0.0       1.0  0.0  0.0  0.0  
2  1.0       0.0  0.0  0.0  0.0  
3  1.0       0.0  1.0  1.0  0.0  
4  0.0       0.0  0.0  0.0  1.0  
5  1.0       0.0  0.0  0.0  0.0

Solución esperada:

|    |   El |   pájaro |   despierta |   el |   come |   a |   nadar |   se |   en |   y |   alimento |   semillas |   pez |   empieza |   canta |   agua |   nada |
|:---|-----:|---------:|------------:|-----:|-------:|----:|--------:|-----:|-----:|----:|-----------:|-----------:|------:|----------:|--------:|-------:|-------:|
| d0 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          0 |          1 |     0 |         0 |       0 |      0 |      0 |
| d1 |    1 |        1 |           1 |    0 |      0 |   0 |       0 |    1 |    0 |   1 |          0 |          0 |     0 |         0 |       1 |      0 |      0 |
| d2 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   1 |          0 |          1 |     0 |         0 |       1 |      0 |      0 |
| d3 |    1 |        0 |           0 |    1 |      1 |   0 |       0 |    0 |    1 |   1 |          0 |          0 |     1 |         0 |       0 |      1 |      1 |
| d4 |    1 |        0 |           0 |    0 |      0 |   1 |       1 |    0 |    0 |   0 |          0 |          0 |     1 |         1 |       0 |      0 |      0 |
| d5 |    1 |        0 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          1 |          0 |     1 |         0 |       0 |      0 |      0 |

### 4.b IDF (0.5 puntos)

Implementar `calc_idf(dataset_bow)`. Ésta debe retornar un diccionario en donde las llaves sean las palabras y los valores sean el cálculo de cada idf por palabra.

Recordar que $\text{idf}_{t_i} = \log_{10}\frac{N}{n_i}$ con $N = $ número de documentos y $n_i = $ número de documentos que contienen la palabra $t_i$

In [131]:
import numpy as np

In [132]:
def calc_idf(dataset_bow):
    
    # Crear un diccionario para almacenar los valores de IDF
    idf = {}

    # Número de documentos
    N = len(dataset_bow)

    # Recorrer columnas
    for i in range(len(dataset_bow.columns)):
        # Calcular la suma de todos los elementos de la columna
        n_i = dataset_bow[dataset_bow.columns[i]].sum()

        # Calcular el IDF para la palabra i
        idf[dataset_bow.columns[i]] = np.log10(N / n_i)

    return idf

In [133]:
idf = calc_idf(dataset_bow)
idf

{'se': 0.7781512503836436,
 'pez': 0.3010299956639812,
 'El': 0.0,
 'pájaro': 0.3010299956639812,
 'en': 0.7781512503836436,
 'nada': 0.7781512503836436,
 'alimento': 0.7781512503836436,
 'y': 0.3010299956639812,
 'nadar': 0.7781512503836436,
 'empieza': 0.7781512503836436,
 'semillas': 0.47712125471966244,
 'canta': 0.47712125471966244,
 'come': 0.17609125905568124,
 'despierta': 0.7781512503836436,
 'el': 0.7781512503836436,
 'agua': 0.7781512503836436,
 'a': 0.7781512503836436}

Solución esperada:
```
{'El': 0.0,
 'pájaro': 0.3010299956639812,
 'despierta': 0.7781512503836436,
 'el': 0.7781512503836436,
 'come': 0.17609125905568124,
 'a': 0.7781512503836436,
 'nadar': 0.7781512503836436,
 'se': 0.7781512503836436,
 'en': 0.7781512503836436,
 'y': 0.3010299956639812,
 'alimento': 0.7781512503836436,
 'semillas': 0.47712125471966244,
 'pez': 0.3010299956639812,
 'empieza': 0.7781512503836436,
 'canta': 0.47712125471966244,
 'agua': 0.7781512503836436,
 'nada': 0.7781512503836436}
 ```

### 4.c TF-IDF (0.25 puntos)
Programe la función `calc_tf_idf(tf, idf)` que entrega el dataframe TF-IDF asociado al dataset que estamos analizando.

In [134]:
def calc_tf_idf(tf, idf):
    
    # Crear un DataFrame para almacenar los valores de TF-IDF
    tf_idf = pd.DataFrame(columns=tf.columns, index=range(len(tf)))

    # Recorrer las filas
    for i in range(len(tf)):
        # Calcular el TF-IDF para cada palabra
        for palabra in tf.columns:
            tf_idf.loc[i, palabra] = tf.loc[i, palabra] * idf[palabra]

    return tf_idf

In [135]:
tf_idf = calc_tf_idf(tf, idf)
tf_idf

se      pez   El   pájaro        en      nada  alimento        y  \
0       0.0      0.0  0.0  0.30103       0.0       0.0       0.0      0.0   
1  0.778151      0.0  0.0  0.30103       0.0       0.0       0.0  0.30103   
2       0.0      0.0  0.0  0.30103       0.0       0.0       0.0  0.30103   
3       0.0  0.30103  0.0      0.0  0.778151  0.778151       0.0  0.30103   
4       0.0  0.30103  0.0      0.0       0.0       0.0       0.0      0.0   
5       0.0  0.30103  0.0      0.0       0.0       0.0  0.778151      0.0   

      nadar   empieza  semillas     canta      come despierta        el  \
0       0.0       0.0  0.477121       0.0  0.176091       0.0       0.0   
1       0.0       0.0       0.0  0.477121       0.0  0.778151       0.0   
2       0.0       0.0  0.477121  0.477121  0.176091       0.0       0.0   
3       0.0       0.0       0.0       0.0  0.176091       0.0  0.778151   
4  0.778151  0.778151       0.0       0.0       0.0       0.0       0.0   
5       0.0       0.0       0.0       0.0  0.176091       0.0       0.0   

       agua         a  
0       0.0       0.0  
1       0.0       0.0  
2       0.0       0.0  
3  0.778151       0.0  
4       0.0  0.778151  
5       0.0       0.0

Solución esperada:

|    |   El |   pájaro |   despierta |       el |     come |        a |    nadar |       se |       en |       y |   alimento |   semillas |     pez |   empieza |    canta |     agua |     nada |
|:---|-----:|---------:|------------:|---------:|---------:|---------:|---------:|---------:|---------:|--------:|-----------:|-----------:|--------:|----------:|---------:|---------:|---------:|
| d0 |    0 |  0.30103 |    0        | 0        | 0.176091 | 0        | 0        | 0        | 0        | 0       |   0        |   0.477121 | 0       |  0        | 0        | 0        | 0        |
| d1 |    0 |  0.30103 |    0.778151 | 0        | 0        | 0        | 0        | 0.778151 | 0        | 0.30103 |   0        |   0        | 0       |  0        | 0.477121 | 0        | 0        |
| d2 |    0 |  0.30103 |    0        | 0        | 0.176091 | 0        | 0        | 0        | 0        | 0.30103 |   0        |   0.477121 | 0       |  0        | 0.477121 | 0        | 0        |
| d3 |    0 |  0       |    0        | 0.778151 | 0.176091 | 0        | 0        | 0        | 0.778151 | 0.30103 |   0        |   0        | 0.30103 |  0        | 0        | 0.778151 | 0.778151 |
| d4 |    0 |  0       |    0        | 0        | 0        | 0.778151 | 0.778151 | 0        | 0        | 0       |   0        |   0        | 0.30103 |  0.778151 | 0        | 0        | 0        |
| d5 |    0 |  0       |    0        | 0        | 0.176091 | 0        | 0        | 0        | 0        | 0       |   0.778151 |   0        | 0.30103 |  0        | 0        | 0        | 0        |

## P5. Cosine-similarity (0.25 puntos)
Ahora que tenemos el dataframe de TF-IDF, nos queda calcular la similitud coseno entre todos los vectores. Notar que la matriz resultante será una matriz simétrica.

Implemente la función *cosine_similarity(v1, v2)* que recibe dos vectores (v1 y v2) y calcula la similitud coseno entre ambos. Concluya cuáles son los dos documentos más similares.

In [136]:
def cosine_similarity(v1, v2):

    # Calcular el producto punto de los vectores
    dot_product = np.dot(v1, v2)

    # Calcular la norma de los vectores
    norm_v1 = np.linalg.norm(v1)
    norm_v2 = np.linalg.norm(v2)

    # Calcular la similitud coseno
    similarity = dot_product / (norm_v1 * norm_v2)

    return similarity

In [137]:
similarity_matrix = np.zeros((6,6))
for i, v1 in enumerate(tf_idf.index.values):
  for j, v2 in enumerate(tf_idf.index.values):
      similarity = cosine_similarity(tf_idf.loc[v1].values, tf_idf.loc[v2].values)
      similarity_matrix[i][j] = similarity

for i in range(6):
  mask = [k != i for k in range(6)]
  j = np.argmax(similarity_matrix[i][mask])

  print(corpus[i])
  print("> Mas similar:", np.array(corpus)[mask][j])
  print("> Similitud:", similarity_matrix[i][mask][j], "\n")

El pájaro come semillas
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.7233435041520413 

El pájaro se despierta y canta
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.3932010182312894 

El pájaro canta y come semillas
> Mas similar: El pájaro come semillas
> Similitud: 0.7233435041520413 

El pez come y nada en el agua
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.09171890791406166 

El pez empieza a nadar
> Mas similar: El pez come alimento
> Similitud: 0.07695078406752713 

El pez come alimento
> Mas similar: El pez come y nada en el agua
> Similitud: 0.0878790037217323 



Solución esperada:
```
El pájaro come semillas
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.7233435041520414

El pájaro se despierta y canta
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.39320101823128945

El pájaro canta y come semillas
> Mas similar: El pájaro come semillas
> Similitud: 0.7233435041520414

El pez come y nada en el agua
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.09171890791406168

El pez empieza a nadar
> Mas similar: El pez come alimento
> Similitud: 0.07695078406752713

El pez come alimento
> Mas similar: El pez come y nada en el agua
> Similitud: 0.0878790037217323
```

## P6 N-gramas (0.75 punto)

En esta sección debera determinar los n-gramas del la cancion "Oh algoritmo".

### 6.a Corpus de entrenamiento y test (0.25 puntos)

En esta subsección debera definir el conjunto de entrenamiento y test de un corpus. Eliga una particion del 80% y 20% del texto.

In [138]:
# Abrir oh_algoritmo.txt
with open('oh_algoritmo.txt', 'r') as file:
    texto = file.read()

Defina una funcion `get_sentences()` que entregue todas las oraciones del corpus que contengan al menos una palabra.

In [140]:
def get_sentences(texto):

  texto_copy = texto
  
  # Reemplazar saltos de línea por puntos
  texto_copy = texto_copy.replace("\n", ".")
  # Separar oraciones
  sentences = texto_copy.split(".")
  # Eliminar oraciones vacías
  sentences = [sentence for sentence in sentences if sentence.strip()]
  # Juntar 2 primeras (Caso Borde)
  sentences = [sentences[0] + sentences[1]] + sentences[2:]
  
  return sentences

In [141]:
oraciones_limpias = get_sentences(texto)
oraciones_limpias

['[Letra de "¡Oh, Algoritmo!" ft Nora Erez]',
 '[Refrán: Jorge Drexler]',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '[Estribillo: Jorge Drexler]',
 'Dime qué debo cantar',
 'Oh, algoritmo',
 'Sé que lo sabes mejor',
 'Incluso que yo mismo',
 '[Verso 1: Nora Erez]',
 "Wait, what's that money that you spent?",
 "What's that sitting on your plate?",
 "Do you want what you've been fed?",
 'Are you the fish or bait?',
 "Mmm, I'm on the top of the roof and I feel like a jail",
 'Rather not pay the bail',
 'To dangerous people with blood on their faces',
 "So I'm sharing a cell with the masses",
 'The underground always strive for the main',
 "Streaming like Grande's big-ass ring",
 "Screaming: I'll wri

Debería obtener en total 87 oraciones.

In [142]:
len(oraciones_limpias) == 87

True

Ahora definiremos el conjunto de entrenamiento y prueba para las oraciones:

In [143]:
split = int(len(oraciones_limpias) * 0.8)
train_corpus = oraciones_limpias[:split]
test_corpus = oraciones_limpias[split:]

len(train_corpus), len(test_corpus)

(69, 18)

### 6.b Estimación de N-gramas (0.5 puntos)

Defina una función que reciba una lista de oraciones de un corpus y un N que indique el tamaño de los N-gramas. La función debe retornar un diccionario de Python donde la llave es un token (o palabra) y el valor es la cantidad de veces que ocurre el token, es decir, la frecuencia. En el caso de N-gramas con N mayor a 1 (como bi-gramas o tri-gramas) debe añadir un token especial al inicio o final de cada oración según corresponda (ver clases del curso).

In [144]:
import nltk
nltk.download('punkt')
from nltk.tokenize import word_tokenize

[nltk_data] Downloading package punkt to
[nltk_data]     /Users/martinbravodiaz/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [145]:
def n_grams(corpus, n=3):
      
      # Crear un diccionario para almacenar los n-gramas
      n_grams = {}
  
      # Recorrer las oraciones del corpus
      for sentence in corpus:
          # Tokenizar la oración
          tokens = word_tokenize(sentence)
          # Recorrer los tokens
          for i in range(len(tokens) - n + 1):
              # Crear el n-grama
              n_gram = " ".join(tokens[i:i+n])
              # Agregar el n-grama al diccionario
              if n_gram in n_grams:
                  n_grams[n_gram] += 1
              else:
                  n_grams[n_gram] = 1
  
      return n_grams

In [146]:
n_grams(train_corpus, n=1)

{'[': 10,
 'Letra': 1,
 'de': 3,
 '``': 1,
 '¡Oh': 1,
 ',': 11,
 'Algoritmo': 1,
 '!': 1,
 "''": 1,
 'ft': 1,
 'Nora': 3,
 'Erez': 3,
 ']': 10,
 'Refrán': 2,
 ':': 11,
 'Jorge': 10,
 'Drexler': 10,
 '¿Quién': 11,
 'quiere': 11,
 'que': 38,
 'yo': 14,
 'quiera': 11,
 'lo': 13,
 'creo': 11,
 'quiero': 11,
 '?': 18,
 'Estribillo': 2,
 'Dime': 3,
 'qué': 3,
 'debo': 3,
 'cantar': 3,
 'Oh': 2,
 'algoritmo': 3,
 'Sé': 2,
 'sabes': 2,
 'mejor': 2,
 'Incluso': 2,
 'mismo': 2,
 'Verso': 4,
 '1': 1,
 'Wait': 1,
 'what': 3,
 "'s": 3,
 'that': 4,
 'money': 1,
 'you': 6,
 'spent': 1,
 'What': 1,
 'sitting': 1,
 'on': 3,
 'your': 1,
 'plate': 1,
 'Do': 1,
 'want': 4,
 "'ve": 1,
 'been': 1,
 'fed': 1,
 'Are': 1,
 'the': 8,
 'fish': 1,
 'or': 1,
 'bait': 1,
 'Mmm': 1,
 'I': 7,
 "'m": 2,
 'top': 1,
 'of': 1,
 'roof': 1,
 'and': 1,
 'feel': 1,
 'like': 3,
 'a': 3,
 'jail': 1,
 'Rather': 1,
 'not': 3,
 'pay': 1,
 'bail': 1,
 'To': 1,
 'dangerous': 1,
 'people': 1,
 'with': 2,
 'blood': 1,
 'their': 1,
 '

In [147]:
n_grams(train_corpus, n=2)

{'[ Letra': 1,
 'Letra de': 1,
 'de ``': 1,
 '`` ¡Oh': 1,
 '¡Oh ,': 1,
 ', Algoritmo': 1,
 'Algoritmo !': 1,
 "! ''": 1,
 "'' ft": 1,
 'ft Nora': 1,
 'Nora Erez': 3,
 'Erez ]': 3,
 '[ Refrán': 2,
 'Refrán :': 2,
 ': Jorge': 7,
 'Jorge Drexler': 10,
 'Drexler ]': 7,
 '¿Quién quiere': 11,
 'quiere que': 11,
 'que yo': 13,
 'yo quiera': 11,
 'quiera lo': 11,
 'lo que': 11,
 'que creo': 11,
 'creo que': 11,
 'que quiero': 11,
 'quiero ?': 11,
 '[ Estribillo': 2,
 'Estribillo :': 2,
 'Dime qué': 3,
 'qué debo': 3,
 'debo cantar': 3,
 'Oh ,': 2,
 ', algoritmo': 2,
 'Sé que': 2,
 'que lo': 2,
 'lo sabes': 2,
 'sabes mejor': 2,
 'Incluso que': 2,
 'yo mismo': 2,
 '[ Verso': 4,
 'Verso 1': 1,
 '1 :': 1,
 ': Nora': 2,
 'Wait ,': 1,
 ', what': 1,
 "what 's": 1,
 "'s that": 2,
 'that money': 1,
 'money that': 1,
 'that you': 1,
 'you spent': 1,
 'spent ?': 1,
 "What 's": 1,
 'that sitting': 1,
 'sitting on': 1,
 'on your': 1,
 'your plate': 1,
 'plate ?': 1,
 'Do you': 1,
 'you want': 2,
 'want wh

In [148]:
n_grams(train_corpus, n=3)

{'[ Letra de': 1,
 'Letra de ``': 1,
 'de `` ¡Oh': 1,
 '`` ¡Oh ,': 1,
 '¡Oh , Algoritmo': 1,
 ', Algoritmo !': 1,
 "Algoritmo ! ''": 1,
 "! '' ft": 1,
 "'' ft Nora": 1,
 'ft Nora Erez': 1,
 'Nora Erez ]': 3,
 '[ Refrán :': 2,
 'Refrán : Jorge': 2,
 ': Jorge Drexler': 7,
 'Jorge Drexler ]': 7,
 '¿Quién quiere que': 11,
 'quiere que yo': 11,
 'que yo quiera': 11,
 'yo quiera lo': 11,
 'quiera lo que': 11,
 'lo que creo': 11,
 'que creo que': 11,
 'creo que quiero': 11,
 'que quiero ?': 11,
 '[ Estribillo :': 2,
 'Estribillo : Jorge': 2,
 'Dime qué debo': 3,
 'qué debo cantar': 3,
 'Oh , algoritmo': 2,
 'Sé que lo': 2,
 'que lo sabes': 2,
 'lo sabes mejor': 2,
 'Incluso que yo': 2,
 'que yo mismo': 2,
 '[ Verso 1': 1,
 'Verso 1 :': 1,
 '1 : Nora': 1,
 ': Nora Erez': 2,
 'Wait , what': 1,
 ", what 's": 1,
 "what 's that": 1,
 "'s that money": 1,
 'that money that': 1,
 'money that you': 1,
 'that you spent': 1,
 'you spent ?': 1,
 "What 's that": 1,
 "'s that sitting": 1,
 'that sitting on

Debe mostrar que su método funciona para $N = 1,2,3$.

## P7. Perplexity (1 punto)

En esta sección evaluarán su modelo de n-gramas y determinarán la probabilidad de oraciones y la perplejidad con un conjunto de test. Recuerde que la perplejidad se define de la siguiente manera:

$$
\text{Perplexity} = 2^{-l} \quad \quad l = \frac{1}{M} \sum_{i=1}^{m} \log p(s_i)
$$

con $m$ el número de oraciones del corpus y $M$ el tamaño del vocabulario.

### 7.a Obtener probabilidades (0.5 puntos)

En esta sección implementará una función que determine la probabilidad de una oración.

Defina una función que reciba una oración, un diccionario con n-gramas y el valor de $n$. La función debe entregar la probabilidad de cualquier oración.

**Hint**: No olvide los posibles casos borde, como palabras fuera del vocabulario.

In [169]:
def get_probability(sentence, n_grams_frequency, n):
  
    # Extraer tokens de la oración
    sentence_tokens = word_tokenize(sentence)
    # Crear n-gramas
    n_grams_sentence = [sentence_tokens[i:i+n] for i in range(len(sentence_tokens) - n + 1)]
    # Convertir n-gramas a cadenas
    n_grams_sentence = [" ".join(n_gram) for n_gram in n_grams_sentence]
    # Calcular la probabilidad
    probability = 0
    for n_gram in n_grams_sentence:
        if n_gram in n_grams_frequency:
            probability += n_grams_frequency[n_gram] / sum(n_grams_frequency.values())

    return probability

Pruebe su función con oraciones frecuentes y comente sus resultados



In [170]:
get_probability("Oh, algoritmo", n_grams(train_corpus, n=2), 2)

0.010126582278481013

### 7.b Perplexity en conjunto de test (0.5 puntos)

En esta sub-sección deberá calcular la perplejidad del corpus de test.

Defina una función que reciba un corpus de test y retorne la perplexity (ver clases del curso). Utilice la función de la sección anterior.

In [171]:
def get_perplexity(corpus, n):

  # Obtener el largo del vocabulario
  vocab = get_vocab(corpus)
  M = len(vocab)

  # Obtener el largo de las oraciones
  m = len(corpus)

  # Calcular la sumatoria 
  sum = 0
  for i in range(m):
    sum += np.log(get_probability(corpus[i], n_grams(corpus, n), n))
  l = sum / M

  # Calcular la perplexity
  perplexity = 2 ** (-l)

  return perplexity

In [172]:
test_corpus

['(Oh, algoritmo)',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '(Sé que lo sabes mejor)',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '(Incluso que yo mismo)',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '(Dime qué debo cantar)',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '(Oh, algoritmo)',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '(Sé que lo sabes mejor)',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '(Incluso que yo mismo)',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '(Wow)']

In [173]:
get_perplexity(test_corpus, 3)

1.2082530160514944

Dé una interpretacion de la perplexity en el corpus de test:
```
Nosotros 
```

## P8. Interpolación Lineal (0.5 puntos)

Cree una función que obtenga la probabilidad de una oración interpolando linealmente modelos de unigrama, bigrama y trigrama ponderados por $\lambda_1, \lambda_2$ y $\lambda_3$ respectivamente. Para esto use las funciones que creó anteriormente.

In [174]:
def get_probability_lineal_interpol(sentence, corpus, l_1, l_2, l_3):
  
    # Calcular la probabilidad
    probability = 0
    for i in range(1, 4):
      probability += get_probability(sentence, n_grams(corpus, i), i) * [l_1, l_2, l_3][i-1]

    return probability

Defina una función para calcular la perplejidad de un corpus con interpolación lineal.

In [175]:
def get_pp_interpol(corpus, l_1, l_2, l_3):
  
    # Obtener el largo del vocabulario
    vocab = get_vocab(corpus)
    M = len(vocab)

    # Obtener el largo de las oraciones
    m = len(corpus)

    # Calcular la sumatoria 
    sum = 0
    for i in range(m):
      sum += np.log(get_probability_lineal_interpol(corpus[i], corpus, l_1, l_2, l_3))
    l = sum / M

    # Calcular la perplexity
    perplexity = 2 ** (-l)

    return perplexity

Ahora haga pruebas con distintos valores de $\lambda_1, \lambda_2$ y $\lambda_3$, incluyendo valores extremos (por ejemplo $[1, 0, 0]$). Comente sus resultados.

In [176]:
get_pp_interpol(test_corpus, 1, 0, 0)

1.0695087602131068